### PR TITLE
Fix for TypeScript preprocessor token type incompatibility

### LINKED ISSUE #527: Type incompatibility between nearly and moo
First of all, thanks for creating this library. Developing with nearley has been such a great experience. 

---

While moo is the recommended lexer for nearley, using moo along with ts postprocessor results in type errors in the generated grammar.

Minimal example: 

```ts
import moo from "moo";

const l = moo.compile({})

interface NearleyLexer {
  reset: (chunk: string, info: any) => void;
  next: () => NearleyToken | undefined;
  save: () => any;
  formatError: (token: NearleyToken) => string;
  has: (tokenType: string) => boolean;
};

interface Grammar {
  Lexer: NearleyLexer | undefined;
  ParserRules: NearleyRule[];
  ParserStart: string;
};

const grammar: Grammar = {
  Lexer: l,
  ParserRules: [
      // ... 
  ]
}
```

This fails with following error: 

```
Type 'Lexer' is not assignable to type 'NearleyLexer'.
  Types of property 'formatError' are incompatible.
    Type '(token: Token, message?: string | undefined) => string' is not assignable to type '(token: NearleyToken) => string'.
      Types of parameters 'token' and 'token' are incompatible.
        Type 'NearleyToken' is missing the following properties from type 'Token': offset, text, lineBreaks, line, col
```

Caveat is that  function type parameter positions are checked contravariantly in strict mode in TS, which breaks this. 

A simple solution is to convert the generated interface to have methods instead of member properties that are functions, in which bivariant type checking is applied: 

```ts
interface NearleyLexer {
    reset(chunk: string, info: any): void;
    next(): NearleyToken | undefined;
    save(): any;
    formatError(token: NearleyToken): string;
    has(tokenType: string): boolean;
  };
```

With the above type in place, the type checker does not complain anymore.

### PR DESCRIPTION
**Pull request by bandaloo**
This is a much less invasive fix that resolves #527

### CODE REVIEW

File: lib/generate.js [hunk 1]
```diff
@@ -56,7 +56,7 @@
         }
 
         options = options || {};
-        tabulated = lines.map(function addIndent(line, i) {
+        var tabulated = lines.map(function addIndent(line, i) {
```
Comment (kach) [inline lib/generate.js:59]: Good catch!

File: lib/generate.js [hunk 2]
```diff
@@ -193,16 +193,11 @@
         output += parser.customTokens.map(function (token) { return "declare var " + token + ": any;\n" }).join("")
         output += parser.body.join('\n');
         output += "\n";
-        output += "interface NearleyToken {";
-        output += "  value: any;\n";
-        output += "  [key: string]: any;\n";
-        output += "};\n";
-        output += "\n";
```
Comment (kach) [inline lib/generate.js:200]: Wait… are you sure this should be deleted? Isn't `NearleyToken` referenced in the type signature for `next()` a couple lines below?

**Comment by kach** [comment]
Thanks for proactively submitting this PR! See my comments.

File: lib/generate.js (hunk 2 above)
Comment (bandaloo) [reply lib/generate.js:200]: um yeah not sure what i was thinking sorry. somehow i ran the wrong test

**Comment by kach** [comment]
Looks good! Last request: Is there a test you can _add_ that passes after this change but fails before this change?

**Comment by bandaloo** [comment]
I added a test that only passes with the `never` change to the generator, which I learned is only needed with the `--strict` compilation mode (but I think that's how most people use TypeScript anyways). In doing so, I also realized that the original typescript test does not fail if there are type errors! That's because `tsc` has no problem handing you a js file even if there were type errors, and the type errors get logged to stdout. I fixed this too. (That's how I made the mistake of lopping off the `NearleyToken` interface and felt confident because the tests still passed, because type errors weren't caught in the tests.)

**Comment by kach** [comment]
Merged! Thanks so much!

Let's watch out for any TS-related issues reported in the next few days… I'm always super paranoid about changes like this. But I think we - really, _you_ - did a great job here. :)

**Pull request merged**